core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala

```diff
@@ -202,6 +214,156 @@ class CompactEntityQuery(driverComponent: DriverComponent) extends RawSqlQuery w
+            and e.entity_type = $entityType
+            and e.workspace_id = $workspaceId
+       """,
+      buildFilterConditions(entityQuery)
+    )
+      .as[Int]
+      .map(_.head)
+
+  private def buildFilterConditions(entityQuery: EntityQuery) = {
+    val filterClauses = entityQuery.filterTermsList.map { filterTerm =>
+      sql"""JSON_SEARCH(e.attributes, 'one', ${'%' + filterTerm.toLowerCase + '%'})"""
+    }
+    concatSqlActions(
+      sql" and (",
+      reduceSqlActionsWithDelim(filterClauses, sql" #${FilterOperators.toSql(entityQuery.filterOperator)} "),
+      sql")"
+    )
+  }
+
+  private def entityResultGetterWithFieldsFilter(entityQuery: EntityQuery) =
+    entityQuery.fields.fields match {
+      case Some(fields) =>
+        // this special GetResult instance is the only way I found to filter the fields and keep this a streaming result
+        val desiredFields = fields.map(AttributeName.fromDelimitedName)
+        val getResultFilteringFields =
+          GetResult(r => CompactEntityRecord(r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, keepOnlyFields(r.<<, desiredFields)))
+        getResultFilteringFields
+      case _ => getJsonEntityRecord
+    }
+
+  def queryEntitiesWithFilterTermsSortByName(workspaceId: UUID,
+                                             entityType: String,
+                                             entityQuery: EntityQuery
+  ): SqlStreamingAction[Seq[CompactEntityRecord], CompactEntityRecord, Read] =
+    concatSqlActions(
+      sql"""#$basicCompactEntitySelect
+            #$fromEntityWhereNotDeleted
+            and e.entity_type = $entityType
+            and e.workspace_id = $workspaceId
+       """,
+      buildFilterConditions(entityQuery),
+      sql""" order by name #${SortDirections.toSql(entityQuery.sortDirection)}
+              limit ${entityQuery.pageSize} offset ${entityQuery.offset}"""
+    ).as[CompactEntityRecord](entityResultGetterWithFieldsFilter(entityQuery))
+
+  def queryEntitiesWithFilterTermsSortByAttribute(workspaceId: UUID,
+                                                  entityType: String,
+                                                  entityQuery: EntityQuery
+  ): SqlStreamingAction[Seq[CompactEntityRecord], CompactEntityRecord, Read] =
+    concatSqlActions(
+      sql"""#$basicCompactEntitySelect,
+            JSON_EXTRACT(e.attributes, ${slickAttributePath(entityQuery.sortField)}) as sort_field,
+            JSON_LENGTH(JSON_EXTRACT(e.attributes, ${slickAttributePath(entityQuery.sortField)})) as sort_length
+            #$fromEntityWhereNotDeleted
+            and e.entity_type = $entityType
+            and e.workspace_id = $workspaceId
+         """,
+      buildFilterConditions(entityQuery),
+      sql""" order by sort_length, sort_field #${SortDirections.toSql(entityQuery.sortDirection)}
+                limit ${entityQuery.pageSize} offset ${entityQuery.offset}"""
+    ).as[CompactEntityRecord](entityResultGetterWithFieldsFilter(entityQuery))
```

Comment:
what is the reasoning to include `sort_length` in the order-by clause? Is the intention to sort scalars before arrays/objects? If so, that seems ok to me but could result in different behavior in legacy vs. Quicksilver tables.

Reply:
I believe this is replicating [legacy behavior](https://github.com/broadinstitute/rawls/blob/8b0baabc17279feb54ef51b05975883485e3f8b5/core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala#L538). My read is that sorting on a list column should sort by the list size and sorting on a scalar column sorts on the column value. If the column is a mixed type then all scalars will group together sorted by value then all the lists will follow sorted by size.

The root of the problem is that we don't know type information before doing the query and we allow mixed types.

Reply:
whoa, TIL! Got it.